### ISSUE TITLE
Альтернативы bemtree

### ISSUE THREAD

**Issue by kompolom** [bem/bem-forum-content-ru#894]
Уже как то заходила речь о там что было бы круто выполнять bemtree в php. Это бы позволило сильно упростить написание шаблонов и код почище бы стал. 

Можете что-то посоветовать, как организовать шаблонизацию не имея bemtree и не усложняя шаблоны блоков?

**Comment by zxqfox** [comment]
Про два набора bh-php (один для тегов, второй для построения дерева) говорить стоит?

**Comment by kompolom** [comment]
@zxqfox ты имеешь ввиду шаблоны или 2 инстанса `$BH`?

**Comment by zxqfox** [comment]
2 инстанса BH и 2 технологии: bh.php и, условно, bh-tree.php (или bh.tree.php), первый набор из bh.php — из BEMJSON собирает HTML чрез `$bh->apply(…)`, второй — как BEMTREE: прокидывает данные (через `$ctx->tParam('data', $data)` например, или еще как-то) и строит BEMJSON через `$bht->processBemjson(['block' => 'root', 'data' => $data])`.

**Comment by kompolom** [comment]
А какие будут ограничения по сравнению с обычным bemtree? 
Когда bemtree был асинхронным и блоки могли ходить за данными. Разница была очевидна. Сейчас все +- одинаково будет работать или я что то упускаю? С bemtree очень мало практики было...

**Comment by zxqfox** [comment]
@kompolom учитывая, что в пхп походы за данными синхронные — разницы не было даже с асинхронным bemtree. ;-)

**Comment by zxqfox** [comment]
> А какие будут ограничения по сравнению с обычным bemtree?

Разница примерно такая же, как между BH и BEMHTML.
В BEMTREE есть произвольный `.match(function() {})`, в BH только по блокам/элементам/модификаторам (решается через разводной матчер, который проставляет модификаторы, на которые отдельные матчеры, и через beforeEach — аналог `block('*')`).
Еще нет подпредикатов на `.tag`, `.bem`, `.cls`, etc. Решается через общий матчер, аналог моды `.def()`.

**Comment by kompolom** [comment]
@zxqfox `$bht->processBemjson()` вернет строку как я понимаю?

**Comment by zxqfox** [comment]
@kompolom нет, оно вернет bemjson. в BH (и BH.PHP) сборка HTML-строки вынесена в отдельную стадию, и это даёт возможность отложить сборку.
Например, можно запускать `bh.processBemjson`, а после результат вставлять в posthtml, потому что там аналогичное дерево, которое, правда, не знает про block/elem/mix/etc, но знает про tag, class, attrs ;-).

**Comment by kompolom** [comment]
@zxqfox я имею ввиду с точки зрения php это будет строка или массив?

**Comment by zxqfox** [comment]
Толи ArrayObject, толи массив из строк и [`\BEM\Json` объектов](https://github.com/bem/bh-php/blob/master/src/Json.php).

**Comment by kompolom** [comment]
@zxqfox тоесть двойного преобразования массив->строка->массив->html не будет?

**Comment by zxqfox** [comment]
Будет такое:
Данные →($bht->processBemjson)→ свернутый `ArrayObject<BEM\Json>` →($bh->processBemjson)→ Отшаблонизированный `ArrayObject<BEM\Json>` →($bh->toHtml)→ HTML-строка.

Данные это ассоц. массив с входными данными типа `{block: 'root', user, data}`.
Свернутый BEMJSON это то, что обычно лежит в .bemjson.js в бандле.
Отшаблонизированный или развернутый BEMJSON это промежуточное вьюориентированное дерево с расставленными tag, attrs, mix, class, и прочим согласно шаблонам, но еще не сконвертированное в строку.
Итоговая HTML-строка — обычный результат, который мы обычно видим в .html файле в бандле.

Каждый этап можно отдебажить, вроде бы не должно быть проблем в реализации в любой CMS.

**Comment by kompolom** [comment]
@zxqfox а как в первый раз у нас получается свернутый bemjson, а когда ту же функцию еще раз применяем развернутый?

**Comment by zxqfox** [comment]
Свернутый-развернутый — это наши условности. processBemjson просто применяет шаблоны, в первом случае это то, что мы положим в $bht, во втором — в $bh (обычные шаблоны).
И мы уславливаемся, что в первом случае мы строим дерево и делаем всякую магию (из глобала берем данные, например), чтобы построить дерево, а во втором — превращаем в HTML (`$bh->apply` это `$bh->toHtml($bh->processBemjson($bemjson))`).

**Comment by kompolom** [comment]
@zxqfox Спасибо! пошел пробовать..
